File: official/atdd.extension.train-interlocking-enforcement/implementations/interlocking-infrastructure/fixtures/clean/python/trains/runner.py
"""TrainRunner — CLEAN fixture.

The production linear train executor: the ONLY component that runs wagons and carries Cargo between
them. This is legitimate execution and is not flagged (the detector only scans InterlockingRunner-
defining modules for direct wagon execution).
"""
from match.wagon import run_match


class Cargo(dict):
    """Artifact data plane carried between wagons inside a selected train."""


class TrainRunner:
    def __init__(self, train_path):
        self._train_path = train_path

    def execute(self, train_id, inputs, timing=None, capture_trace=True):
        cargo = Cargo(inputs)
        train = _load_train(self._train_path)
        for step in train.sequence:
            cargo = run_train(step, cargo)
        return {"selected_train_id": train_id, "artifacts": dict(cargo)}


def run_train(step, cargo):
    if step == "match":
        return run_match(cargo)
    return cargo


def _load_train(path):
    class _T:
        sequence = ["match"]

    return _T()

File: official/convex.workspace.runtime/implementations/convex_interlocking_infrastructure/fixtures/dirty/interlocking_runner_exists/convex/trains/runner.ts
// TrainRunner — CLEAN fixture. The production linear train executor: the ONLY component that runs
// wagons and carries Cargo between them. Legitimate execution; not scanned by the interlocking
// detector (only InterlockingRunner-defining modules are scanned for direct wagon execution).
import { runMatch } from "../match/wagon";

export class Cargo extends Map<string, unknown> {}

export class TrainRunner {
  constructor(private readonly trainPath: string) {}

  execute(trainId: string, inputs: Record<string, unknown>) {
    let cargo: Cargo = new Cargo(Object.entries(inputs));
    const train = loadTrain(this.trainPath);
    for (const step of train.sequence) {
      cargo = runTrain(step, cargo);
    }
    return { selectedTrainId: trainId, artifacts: Object.fromEntries(cargo) };
  }
}

function runTrain(step: string, cargo: Cargo): Cargo {
  if (step === "match") return runMatch(cargo) as Cargo;
  return cargo;
}

function loadTrain(_path: string) {
  return { sequence: ["match"] };
}
